pkg/scheduler/scheduler.go

```diff
@@ -252,12 +278,35 @@ func filterWorkersByPoolSelector(workers []*types.Worker, request *types.Contain
 }
 
 func filterWorkersByResources(workers []*types.Worker, request *types.ContainerRequest) []*types.Worker {
+	gpuRequestsMap := map[string]int{}
+
+	for index, gpu := range request.GpuRequest {
+		gpuRequestsMap[gpu] = index
+	}
+
 	filteredWorkers := []*types.Worker{}
 	for _, worker := range workers {
-		if worker.FreeCpu >= int64(request.Cpu) && worker.FreeMemory >= int64(request.Memory) &&
-			worker.Gpu == request.Gpu && worker.FreeGpuCount >= request.GpuCount && worker.Status != types.WorkerStatusDisabled {
-			filteredWorkers = append(filteredWorkers, worker)
+		// Check if the worker has enough free cpu and memory to run the container
+		if worker.FreeCpu < int64(request.Cpu) || worker.FreeMemory < int64(request.Memory) {
+			continue
+		}
+
+		// Check if the worker has been cordoned
+		if worker.Status == types.WorkerStatusDisabled {
+			continue
+		}
+
+		if len(gpuRequestsMap) > 0 {
+			// Validate GPU resource availability
+			priorityModifier, validGpu := gpuRequestsMap[worker.Gpu]
+			if !validGpu || worker.FreeGpuCount < request.GpuCount {
+				continue
+			}
+
+			worker.Priority -= int32(priorityModifier) // This will account for the preset priorities for the pool type as well as the order of the GPU requests
```

Comment:
hm interesting. will this always work? i think it'll only work if priorities are just increasing by 1?

Reply:
right for now it will work well if priorities are incremental. Once we think of more complex priorities, we would need to update this scoring function. But this additive/subtractive behavior works well with the default priorities.